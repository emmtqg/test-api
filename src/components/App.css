.browsehappy {
  margin: 0.2em 0;
  background: #CCCCCC;
  color: black;
  padding: 0.2em 0;
}

body {
  padding-top: 20px;
  padding-bottom: 20px;
}

.header,
.footer {
  padding-left: 15px;
  padding-right: 15px;
}

.header {
  border-bottom: 1px solid #E5E5E5;
}

.header h3 {
  margin-top: 0;
  margin-bottom: 0;
  line-height: 40px; /* Masthead heading the same height as the navigation */
  padding-bottom: 19px;
}

main {
  margin: 0 auto;
  max-width: 740px;
}

/* Custom page footer */

.footer {
  padding-top: 19px;
  color: #777777;
  border-top: 1px solid #E5E5E5;
}

.container-narrow > hr {
  margin: 30px 0;
}

.splash {
  text-align: center;
  margin: 100px auto 10px;
  border-bottom: 1px solid #E5E5E5;
  background: transparent url('../assets/office_xlarge.jpg') no-repeat center center;
  max-width: 90%;
  height: 400px;
}

.splash h6,
.splash h5 {
  padding: 1em 0;
  color: white;
  background: rgba(0%, 0%, 0%, 0.2);
}

@media screen and ( min-width: 768px ) {

  .container {
    max-width: 100%;
  }

  /* Remove the padding set earlier */

  .header,
  .footer {
    padding-left: 0;
    padding-right: 0;
  }

  .header {
    margin-bottom: 30px;
  }
}
