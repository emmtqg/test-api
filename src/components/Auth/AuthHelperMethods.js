import decode from 'jwt-decode';

import config from '../Config';

// The apiCamelizedLookup provides a fieldname check and crossreference
// for the form field names. It aids in translation of valid property
// names to whatever the API required (single point of definition between
// the state properties and api names). It was assumed if the ESLint rules
// did not allow underscores (as specified in the API names) then, rather
// than overriding the linter, a mechanism should be incorporated to allow
// any API to specify properties as required and they can then be updated
// in one location. If form fields are checked for testing, then the form
// component can also utilize the apiCamelizedLookup for field specifications.
const { apiCamelizedLookup } = config;
const apiNames = Object.keys(apiCamelizedLookup);

export default class AuthHelperMethods {
  getApiPropName = (name) => {
    // Get the API 'expected' property of input form field
    let index = -1;
    Object.values(apiCamelizedLookup).findIndex((ele, i) => {
      index = i;
      return (ele === name);
    });
    return index; // returning matched index or -1 if not in apiCamelized object
  }

  lookupApiName = (name) => {
    const index = this.getApiPropName(name);
    if (index > -1) {
      // apiNames are the API keys in lookup array
      return apiNames[index];
    }

    return ''; // API will error with invalid fieldname
  }

  register = (registerObject) => {
    const {
      firstName,
      lastName,
      email,
      password,
      confirmPassword,
    } = registerObject;

    // Get a token from api server using the fetch api
    return (
      this.fetch(config.registerUrl, {
        method: 'POST',
        body: JSON.stringify({
          [this.lookupApiName('firstName')]: firstName,
          [this.lookupApiName('lastName')]: lastName,
          email,
          password,
          [this.lookupApiName('confirmPassword')]: confirmPassword,
        }),
      })
        .then((response) => {
          // console.log(`Auth (Register): ${response}`);
          if (response) {
            this.setToken(response.token); // Setting the token in localStorage
            return Promise.resolve(response);
          }
          return Promise.reject(response);
        })
    );
  }

  newPlayer = (newPlayerObject) => {
    const {
      firstName,
      lastName,
      rating,
      handedness,
    } = newPlayerObject;

    // Get a token from api server using the fetch api
    return (
      this.fetch(config.playersUrl, {
        method: 'POST',
        body: JSON.stringify({
          [this.lookupApiName('firstName')]: firstName,
          [this.lookupApiName('lastName')]: lastName,
          rating,
          handedness,
        }),
      })
    );
  }

  getPlayers = () => (
    this.fetch(config.playersUrl, {
      method: 'GET',
    })
  );

  removePlayer = (playerId) => {
    const deleteUrl = `${config.playersUrl}/${playerId}`;
    return (
      this.fetch(deleteUrl, {
        method: 'DELETE',
      })
    );
  }

  login = (loginObject) => {
    const {
      email,
      password,
    } = loginObject;

    // Get a token from api server using the fetch api
    return (
      this.fetch(config.loginUrl, {
        method: 'POST',
        body: JSON.stringify({
          email,
          password,
        }),
      })
        .then((response) => {
          if (response) {
            // console.log(`Auth (Login): ${response}`);
            this.setToken(response.token); // Setting the token in localStorage
            return response;
          }
          return Promise.reject(response);
        })
        .catch((error) => {
          if (error && error.message) {
            // console.log(`Error: ${error.message}`);
          }
          return Promise.reject(error);
        })
    );
  }

  loggedIn = () => {
    // Checks if there is a saved token and it's still valid
    const token = this.getToken(); // Getting token from localstorage
    return (typeof token !== typeof undefined) && (!!token && !this.isTokenExpired(token));
  }

  isTokenExpired = (token) => {
    try {
      const decoded = decode(token);
      if (decoded.exp < Date.now() / 1000) { // Checking if token is expired.
        return true;
      }
      return false;
    } catch (err) {
      // console.log('Check for expired tolen failed.');
      return false;
    }
  }

  setToken = (idToken) => {
    // Saves user token to localStorage
    localStorage.setItem('id_token', idToken);
  }

  getToken = () =>
    // Retrieves the user token from localStorage
    localStorage.getItem('id_token')


  logout = () => {
    // Clear user token and profile data from localStorage
    localStorage.removeItem('id_token');
  }

  getConfirm = () => {
    // Using jwt-decode npm package to decode the token
    const answer = decode(this.getToken());
    // console.log(`Get confirm token => ${answer}`);
    return answer;
  }

  fetch = (url, options) => {
    // performs api calls sending the required authentication headers
    const headers = {
      Accept: 'application/json',
      'Content-Type': 'application/json',
    };

    // Setting Authorization header
    // Authorization: Bearer xxxxxxx.xxxxxxxx.xxxxxx
    if (this.loggedIn()) {
      headers.Authorization = `Bearer ${this.getToken()}`;
    }

    return fetch(url, {
      headers,
      ...options,
    })
      .then(this.checkStatus)
      .then(response => response.json())
      .catch((error) => {
        // console.log(`Error: ${error.message}`);
        throw error;
      });
  }

  checkStatus = (response) => {
    // Success status lies between 200 to 300
    if (response.status >= 200 && response.status < 300) {
      return response;
    }

    // raises an error in case response status is not a success
    const error = new Error(response.statusText);
    error.response = response;
    error.status = response.status;
    throw error;
  }
}
