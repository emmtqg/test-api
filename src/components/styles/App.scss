// app.scss

.browsehappy {
  margin: 0.2em 0;
  background: #CCCCCC;
  color: black;
  padding: 0.2em 0;
}

body {
  padding-top: 2rem;
  padding-bottom: 2rem;
}

main {
  margin: 0 auto;
  max-width: 75rem;
}

.header,
.footer {
  padding-left: 1.5rem;
  padding-right: 1.5rem;
}

.header {
  border-bottom: 1px solid #E5E5E5;

  h3 {
    margin-top: 0;
    margin-bottom: 0;
    line-height: 4rem;
    padding-bottom: 2rem;
  }
}

.splash {
  text-align: center;
  margin: 10rem auto 1rem;
  border-bottom: 1px solid #E5E5E5;
  background: transparent url('../../assets/office_xlarge.jpg') no-repeat center center;
  max-width: 90%;
  height: 40rem;

  h5,
  h6 {
    padding: 1em 0;
    color: white;
    background: rgba(0%, 0%, 0%, 0.2);
  }
}

@media screen and ( min-width: 77rem ) {

  .container {
    max-width: 100%;
  }

  .header,
  .footer {
    padding-left: 0;
    padding-right: 0;
  }

  .header {
    margin-bottom: 3rem;
  }
}
