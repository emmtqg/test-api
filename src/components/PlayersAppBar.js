import React, { Component } from 'react';
import PropTypes from 'prop-types';
import { withStyles } from '@material-ui/core/styles';
import AppBar from '@material-ui/core/AppBar';
import Toolbar from '@material-ui/core/Toolbar';
import Button from '@material-ui/core/Button';

import logo from '../assets/logo-alchemy-white.svg';

const styles = {
  root: {
    flexGrow: 1,
  },
  grow: {
    flexGrow: 1,
  },
  menuButton: {
    marginLeft: -12,
    marginRight: 20,
  },
  left: {
    textAlign: 'left',
  },
  right: {
    float: 'right',
  },
  menuButtonRight: {
    float: 'right',
    marginLeft: 10,
  },
};

class PlayersAppBar extends Component {
  constructor(props) {
    super(props);

    this.goTo = this.goTo.bind(this);
    this.logout = this.logout.bind(this);
  }

  goTo(route) {
    this.props.history.replace(`/${route}`);
  }

  logout() {
    this.props.auth.logout();
    this.goTo('');
  }

  render() {
    const { classes, auth } = this.props;
    const isAuthenticated = auth.loggedIn();

    return (
      <div className={classes.root}>
        <AppBar
          position="fixed"
          style={{ background: '#3f51b5' }}
        >
          <Toolbar>
            <img src={logo} alt="Alchemy Logo" height="50" width="136" />
            {
              !isAuthenticated && (
                <div style={{ float: 'right', width: '100%' }}>
                  <Button
                    id="Login"
                    color="secondary"
                    variant="contained"
                    className={classes.menuButtonRight}
                    onClick={() => this.goTo('login')}
                  >
                    Login
                  </Button>
                  <Button
                    id="Register"
                    color="secondary"
                    variant="contained"
                    className={classes.menuButtonRight}
                    onClick={() => this.goTo('register')}
                  >
                    Register
                  </Button>
                </div>
              )
            }
            {
              isAuthenticated && (
                <div style={{ width: '100%' }}>
                  <div style={{ float: 'right' }}>
                    <Button
                      color="secondary"
                      variant="contained"
                      className={classes.menuButtonRight}
                      onClick={this.logout}
                    >
                      Logout
                    </Button>
                  </div>
                </div>
              )
            }
          </Toolbar>
        </AppBar>
      </div>
    );
  }
}

PlayersAppBar.defaultProps = {
  auth: {},
  history: () => {
    this.setState({
      dialogOpen: true,
      dialogMessage: 'No history object passed to this page [Roster]',
    });
  },
};

PlayersAppBar.propTypes = {
  // Material-UI needs to see this defined as object -
  // the classes get wrapped with the styles via withStyles HOC.
  // eslint-disable-next-line react/forbid-prop-types
  classes: PropTypes.object.isRequired,
  auth: PropTypes.shape({
    loggedIn: PropTypes.func.isRequired,
    logout: PropTypes.func.isRequired,
  }),
  // History object passed via Router or a function
  // that passed in unit tests via jest.
  history: PropTypes.oneOfType([
    PropTypes.shape({
      replace: PropTypes.func,
    }),
    PropTypes.func,
  ]),
};

export default withStyles(styles)(PlayersAppBar);
