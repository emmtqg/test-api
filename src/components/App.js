// @flow
import React from 'react';
import AuthHelperMethods from './Auth/AuthHelperMethods';
import PlayersAppBar from './PlayersAppBar';
import './App.css';

const auth = new AuthHelperMethods();

// Props are route { auth, history, location, match } from BrowserRouter
const App = props => <PlayersAppBar {...props} auth={auth} />;

export default App;
