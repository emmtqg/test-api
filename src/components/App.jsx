// @flow
import React from 'react';

import AuthHelperMethods from './auth/AuthHelperMethods';
import PlayersAppBar from './PlayersAppBar';
import './styles/App.scss';

const auth = new AuthHelperMethods();

// Props are route { auth, history, location, match } from BrowserRouter
const App = props => <PlayersAppBar {...props} auth={auth} />;

export default App;
