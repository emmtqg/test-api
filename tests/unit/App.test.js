import React from 'react';
import renderer from 'react-test-renderer';
import { shallow } from 'enzyme';
import App from '../../src/components/App';
import PlayersAppBar from '../../src/components/PlayersAppBar';
import AuthHelperMethods from '../../src/components/auth/AuthHelperMethods';

let wrapper;
const auth = new AuthHelperMethods();

describe('The App', () => {
  beforeEach(() => {
    wrapper = shallow(<App />);
  });

  // smoke test
  it('Smoke test: Should render without crashing', () => {
    shallow(<App />);
  });

  it('shows a header area (MUI AppBar)', () => {
    expect(wrapper.find(PlayersAppBar).length).toEqual(1);
  });

  it('should render correctly', () => {
    const component = renderer.create(<App auth={auth} />);
    expect(component.toJSON()).toMatchSnapshot();
  });
});
