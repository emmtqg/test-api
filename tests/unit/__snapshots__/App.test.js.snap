// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`The App should render correctly 1`] = `
<div
  className="PlayersAppBar-root-1"
>
  <header
    className="MuiPaper-root-16 MuiPaper-elevation4-22 MuiAppBar-root-7 MuiAppBar-positionFixed-8 MuiAppBar-colorPrimary-14 mui-fixed"
    style={
      Object {
        "background": "#3f51b5",
      }
    }
  >
    <div
      className="MuiToolbar-root-43 MuiToolbar-regular-45 MuiToolbar-gutters-44"
    >
      <img
        alt="Alchemy Logo"
        height="50"
        src="test-file-stub"
        width="136"
      />
      <div
        style={
          Object {
            "float": "right",
            "width": "100%",
          }
        }
      >
        <button
          className="MuiButtonBase-root-73 MuiButton-root-47 MuiButton-contained-58 MuiButton-containedSecondary-60 MuiButton-raised-61 MuiButton-raisedSecondary-63 PlayersAppBar-menuButtonRight-6"
          disabled={false}
          id="Login"
          onBlur={[Function]}
          onClick={[Function]}
          onContextMenu={[Function]}
          onFocus={[Function]}
          onKeyDown={[Function]}
          onKeyUp={[Function]}
          onMouseDown={[Function]}
          onMouseLeave={[Function]}
          onMouseUp={[Function]}
          onTouchEnd={[Function]}
          onTouchMove={[Function]}
          onTouchStart={[Function]}
          tabIndex="0"
          type="button"
        >
          <span
            className="MuiButton-label-48"
          >
            Login
          </span>
          <span
            className="MuiTouchRipple-root-76"
          />
        </button>
        <button
          className="MuiButtonBase-root-73 MuiButton-root-47 MuiButton-contained-58 MuiButton-containedSecondary-60 MuiButton-raised-61 MuiButton-raisedSecondary-63 PlayersAppBar-menuButtonRight-6"
          disabled={false}
          id="Register"
          onBlur={[Function]}
          onClick={[Function]}
          onContextMenu={[Function]}
          onFocus={[Function]}
          onKeyDown={[Function]}
          onKeyUp={[Function]}
          onMouseDown={[Function]}
          onMouseLeave={[Function]}
          onMouseUp={[Function]}
          onTouchEnd={[Function]}
          onTouchMove={[Function]}
          onTouchStart={[Function]}
          tabIndex="0"
          type="button"
        >
          <span
            className="MuiButton-label-48"
          >
            Register
          </span>
          <span
            className="MuiTouchRipple-root-76"
          />
        </button>
      </div>
    </div>
  </header>
</div>
`;
